### DISCUSSION TITLE
Any way to see which instrument we are on when sending Program Change command?

### DISCUSSION THREAD

**Discussion by matheustavaresdev** [GeekFunkLabs/fluidpatcher#76] [General]
Hi! First of all thanks for your work! This has been working seamlessly since install! 

But, before using your app, I was using fluidsynth only, and I would load a soundfont, and would use the Program Change command (type=`Program Change` | Channel=1 | Data = 0-127) to change instruments of a soundfont (I guess that's how it's called, I've just bought my first keyboard and it's a midi one so not very used to it yet!). With fluidsynth, I was able to run a command on it's terminal, and see all the list of instruments there (command for that on fluidsynth:`inst 1`). But when running `./headlesspi.py`, I'm only able to see which Patch I'm changing, which means that if I just want to load one font and change between it's instruments, I'd have to create many many patches. 

I think that in the future, I will take notes of the best sounds on a soundfont and create a bank just for them (Unless, in the end, this is the real goal of the code) but, as of now, I only have the LCD display on my keyboard (Nektar LX61+) and it will only give me some direction when ussing the Program Change command... It will display a number from 0-127. I'm planning on buying an LCD display for the rpi, so I can show exactly what I'm doing or which patch at least it's on, but until that, I can only rely on those numbers from 0-127... So I was wondering if there were any way to see this... The instrument I'm changing to, or if what I'm doing should be considered a "hack" and its not the proposal of this application (that might be the case, still getting around it haha)

I also saw another [discussion](https://github.com/albedozero/fluidpatcher/discussions/3#discussioncomment-640829) to run `./headlesspi.py --interactive` but this doesn't seem to work anymore, not sure if it would give more info about what's going on, or if there's some sort of verbose argument to be passed...

**Comment by albedozero** [comment]
The general idea behind this software is - as you say - to create banks that have the best presets from each soundfont you want to use. Or maybe the presets that are most useful for a given performance, musical style, etc. Of course if you find a way to hack/work around it that suits your particular purpose then go for it! A few bits of info that might help:

- By default, program change messages are still passed through to fluidsynth, so if each patch selects a different soundfont on channel 1, you can then use program change messages to scroll through the different presets in the soundfont (although you might need bank select messages as well to get percussion, etc.)
- The only command-line argument to headlesspi.py is an alternative config file. The `--interactive` argument was a bit of a misnomer - it just made the program print more info, and I decided there was no reason for it not to do that always. It prints out the name of the bank file being loaded, the patch name and number/total # patches, and any error messages. You could have it print out a list of the patches in the bank by adding a `print('\n'.join(pxr.patches))` [here](https://github.com/albedozero/fluidpatcher/blob/c1079b9e4709955113641bf0ff96bd4ab40df971/headlesspi.py#L132)
- You might find some help adding an LCD and some buttons by looking at the info for the [SquishBox](https://geekfunklabs.com/products/squishbox/) - you can find schematics, manuals, etc. at the link. The squishbox.py script may be of use to you as well
- If you're into python programming, you could use some of the [API functions](https://github.com/albedozero/fluidpatcher/blob/master/patcher/README.md) to create a version of headlesspi.py that just scrolls through every preset on all available soundfonts. Check out the functions _load_soundfont_ and _select_sfpreset_

**Reply by matheustavaresdev** [reply]
Uhhhhhhhh awesome!!! Nice then! Got the idea! Yea! For me, for now, only using a single soundfont and scroll through it's instruments will work, but I was really thinking in the future and your solution works just fine for what I was thinking that I could need in the long run! Yea, I have quite some programming knowledge, not much of python, but what I have will work for what I will need for this! :D 

Thanks a lot for the tips! I will try to make some changes on it on my free time over weekends! :D If I make any relevant upgrade I will bring it up to see if makes sense to add in the main project! I'm currently low on budget so I'm just buying a LCD display and gonna attach to the rpi. Was already taking a look on the Squishbox code, just not sure if it's gonna work with the current implementation due to driver difference, but I will definitely try it out!
